src/daft-dsl/src/expr.rs

```diff
@@ -1029,6 +1037,197 @@ impl Operator {
     }
 }
 
+/// Converts an expression with syntactic sugar into struct gets.
+/// Does left-associative parsing to to resolve ambiguity.
+///
+/// For example, if col("a.b.c") could be interpreted as either col("a.b").struct.get("c")
+/// or col("a").struct.get("b.c"), this function will resolve it to col("a.b").struct.get("c").
+fn substitute_expr_getter_sugar(expr: ExprRef, schema: &Schema) -> DaftResult<ExprRef> {
+    use common_treenode::{Transformed, TransformedResult, TreeNode};
+
+    #[derive(PartialEq, Eq)]
+    struct BfsState<'a> {
+        name: String,
+        expr: ExprRef,
+        field: &'a Field,
+    }
+
+    impl Ord for BfsState<'_> {
+        fn cmp(&self, other: &Self) -> Ordering {
+            self.name.cmp(&other.name)
+        }
+    }
+
+    impl PartialOrd for BfsState<'_> {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+
+    let mut pq: BinaryHeap<BfsState> = BinaryHeap::new();
```

Comment:
Is the construction of `pq` expensive? Especially if we have to do it for every single logical plan node creation...

An alternative (cheaper but probably more complex) is maybe to perform a walk down the schema for every `Expr::Column`, checking for prefix-matches along the way. This would scale in terms of O(num_col_exprs) instead of O(num_fields_in_schema).

Reply:
Binary heap construction is `O(nlogn)` so it's a little expensive but not too much. Earlier I wrote the solution you mentioned, but that's what led to the messy tree traversal code that I had at the start. There might be a better way to do it but either way you need to traverse both the column name and the schema at the same time which is most of the complexity.